packages/rxjs/src/internal/Observable.ts

```diff
@@ -466,3 +1008,375 @@ export class Observable<T> implements Subscribable<T> {
+ * `from` conditionals because we *know* they're dealing with an array.
+ * @param array The array to emit values from
+ */
+export function fromArrayLike<T>(array: ArrayLike<T>) {
+  return new Observable((subscriber: Subscriber<T>) => {
+    subscribeToArray(array, subscriber);
+  });
+}
+
+export function fromPromise<T>(promise: PromiseLike<T>) {
+  return new Observable((subscriber: Subscriber<T>) => {
+    promise
+      .then(
+        (value) => {
+          if (!subscriber.closed) {
+            subscriber.next(value);
+            subscriber.complete();
+          }
+        },
+        (err: any) => subscriber.error(err)
+      )
+      .then(null, reportUnhandledError);
+  });
+}
+
+function fromIterable<T>(iterable: Iterable<T>) {
+  return new Observable((subscriber: Subscriber<T>) => {
+    for (const value of iterable) {
+      subscriber.next(value);
+      if (subscriber.closed) {
+        return;
+      }
+    }
+    subscriber.complete();
+  });
+}
+
+function fromAsyncIterable<T>(asyncIterable: AsyncIterable<T>) {
+  return new Observable((subscriber: Subscriber<T>) => {
+    process(asyncIterable, subscriber).catch((err) => subscriber.error(err));
+  });
+}
+
+function fromReadableStreamLike<T>(readableStream: ReadableStreamLike<T>) {
+  return fromAsyncIterable(readableStreamLikeToAsyncGenerator(readableStream));
+}
+
+async function process<T>(asyncIterable: AsyncIterable<T>, subscriber: Subscriber<T>) {
+  for await (const value of asyncIterable) {
+    subscriber.next(value);
+    // A side-effect may have closed our subscriber,
+    // check before the next iteration.
+    if (subscriber.closed) {
+      return;
+    }
+  }
+  subscriber.complete();
+}
+
+/**
+ * Subscribes to an ArrayLike with a subscriber
+ * @param array The array or array-like to subscribe to
+ * @param subscriber
+ */
+export function subscribeToArray<T>(array: ArrayLike<T>, subscriber: Subscriber<T>) {
+  // Loop over the array and emit each value. Note two things here:
+  // 1. We're making sure that the subscriber is not closed on each loop.
+  //    This is so we don't continue looping over a very large array after
+  //    something like a `take`, `takeWhile`, or other synchronous unsubscription
+  //    has already unsubscribed.
+  // 2. In this form, reentrant code can alter that array we're looping over.
+  //    This is a known issue, but considered an edge case. The alternative would
+  //    be to copy the array before executing the loop, but this has
+  //    performance implications.
+  const length = array.length;
+  for (let i = 0; i < length; i++) {
+    if (subscriber.closed) {
+      return;
+    }
+    subscriber.next(array[i]);
+  }
+  subscriber.complete();
+}
+
+export enum ObservableInputType {
+  Own,
+  InteropObservable,
+  ArrayLike,
+  Promise,
+  AsyncIterable,
+  Iterable,
+  ReadableStreamLike,
+}
+
+export function getObservableInputType(input: unknown): ObservableInputType {
```

Comment:
NEW: I've centralized the checking of `ObservableInput<T>` here. We had it in two places before, and order mattered. We also had to duplicate the error throwing behavior. This is now used in both `from` and `schedule`.